url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/utility/DataFormater.java

```diff
@@ -0,0 +1,116 @@
+package utility;
+
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.LineNumberReader;
+import java.io.PrintWriter;
+import java.util.HashSet;
+
+public class DataFormater {
+	public static boolean ISCAR = true;
+	public static long SCALE = 100000;
+	public static int DATASET_SIZE = 3000000;
+	public static String input_file_name = "source/NE/NE.txt";
+	public static String output_file_name = "source/NE.dat";
+	
+	public static void genCa(){
+		try{
+			String line = null;
+			LineNumberReader lr = new LineNumberReader(new FileReader(input_file_name));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *lr*. The resource is referenced at lines 21, 26. The resource is closed at line 49. There are other execution paths that do not contain closure statements, for example, when FileWriter constructor throws an exception. Ensure *lr* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).